### ISSUE TITLE
SDK can crash from time to time when using NextJS

### ISSUE THREAD

**Issue by thecoorum** [Telegram-Mini-Apps/issues#18]
## Platform

Select platforms where this bug is reproducible:

[//]: # (To make as selected, use [x] instead of [ ].)

- [ ] Telegram for Android
- [x] Telegram for iOS
- [ ] Telegram for macOS
- [ ] Telegram Desktop
- [ ] Telegram Web A
- [ ] Telegram Web K

## Describe the bug

When using `@tma.js/sdk-react` with NextJS an issue with SDK crashing with `undefined is not an object (evaluating 'window.TelegramGameProxy.receiveEvent')` can appear from time to time

## To Reproduce

1. Mount the SDK provider with the following code
```tsx
"use client";

import { useState, useEffect, useMemo } from "react";

import { useSDKContext } from "@tma.js/sdk-react";

export const ComponentsProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const [render, setRender] = useState(false);
  const { loading, error } = useSDKContext();

  useEffect(() => {
    setRender(true);
  }, []);

  const errorMessage = useMemo<null | string>(() => {
    if (!error) {
      return null;
    }

    return error instanceof Error ? error.message : "Unknown error";
  }, [error]);

  if (!render) return null;

  if (error) {
    setTimeout(() => {
      throw new Error(`SDK error: ${errorMessage}`);
    });

    return <Alert />;
  }

  if (loading) {
    return <Loading />;
  }

  return children;
};
```
2. Create several pages linked to each other with `useRouter` from `next/navigation`
3. Navigate between pages and see the issue from time to time

**Comment by heyqbnk** [comment]
Check if you are using SDKProvider in async mode:

```jsx
<SDKProvider options={{ async: true }}>
....
</SDKProvider>
```

https://docs.telegram-mini-apps.com/packages/typescript/tma-js-sdk-react#asynchronous

**Comment by thecoorum** [comment]
Nope, I'm not (if it's not the default behavior)

**Comment by heyqbnk** [comment]
Also, check out [this](https://github.com/Telegram-Mini-Apps/nextjs-template/blob/master/components/TmaSDKLoader.tsx) component from the template for Next JS which is being used for async mode

**Comment by heyqbnk** [comment]
Seems like I have to update the [template for Next JS](https://github.com/Telegram-Mini-Apps/nextjs-template) also. TBD

**Comment by thecoorum** [comment]
Ah, so if I will switch to `async: true`, can I assume that the `loading` from `useSDKContext` will be `true` until it initialized?

**Comment by heyqbnk** [comment]
> Ah, so if I will switch to `async: true`, can I assume that the `loading` from `useSDKContext` will be `true` until it initialized?

Yes

**Comment by thecoorum** [comment]
Thanks! I will try this approach and get back to you as soon as possible!

**Comment by thecoorum** [comment]
I can see that using `async: true` never resolves `loading` to `true`, so all the children are rendered immediately causing client-side error as soon as SDK is attempted to be queried

**Comment by heyqbnk** [comment]
Yeah, it is not enough just to specify `async: true`. Let me explain.

We created async mode for those developers, who prefer using SSR or who doesn't want to handle async operations at all (bu this approach, ofcourse, has some disadvantages). Using async mode, initially we have such type of SDK state:

```typescript
{
  loading: false
}
```

When component was mounted, it will call initialization, and the state will become the following:
```typescript
{
  loading: true
}
```

In case, some error occurred, then we would have something like that:
```typescript
{
  loading: false,
  error: Error("...")
}
```

Or if everything is fine, the state will be:
```typescript
{
  loading: false,
  initResult: { ... }
}
```

So just checking `error` and `loading` properties from the `useSDKContext` is not enough. You should also check if `initResult` is set. Why? Because initially SDK is not loading, it didn't receive any error and didn't obtain any initialization result. It will start loading only when `SDKProvider` will be mounted to DOM.

According to this, just check all of 3 properties. You can find the example in [docs](https://docs.telegram-mini-apps.com/packages/typescript/tma-js-sdk-react#getting-sdk-context)

**Comment by thecoorum** [comment]
Even with all 3 properties being checked I can still encounter this issue in rare cases

**Comment by heyqbnk** [comment]
Could PM me in Telegram? I think, this problem requires deeper research. https://t.me/vdkfrost

**Comment by thecoorum** [comment]
Sorry, I'm out from laptop for today and don't really think that you would like to dig into this during weekends, so maybe we could discuss it next week? Anyway just say whenever you are comfortable to do this
